### PR TITLE
Person service API client

### PR DESCRIPTION
**Pull request by axthosarouris**

### CODE REVIEW

File: publication-commons/src/main/java/no/unit/nva/publication/external/services/PersonApiClient.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import nva.commons.apigateway.exceptions.ApiGatewayException;
+import nva.commons.apigateway.exceptions.BadGatewayException;
+import nva.commons.core.JacocoGenerated;
+import nva.commons.core.paths.UriWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.zalando.problem.Problem;
+import org.zalando.problem.Status;
+
+public class PersonApiClient {
```
Comment (axthosarouris) [inline publication-commons/src/main/java/no/unit/nva/publication/external/services/PersonApiClient.java:32]: Client that now talks with Bare

File: publication-commons/src/main/java/no/unit/nva/publication/external/services/PersonApiClient.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    public PersonApiClient(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    @JacocoGenerated
+    public PersonApiClient() {
+        this.httpClient = HTTP_CLIENT;
+    }
+
+    public List<URI> fetchAffiliationsForUser(String feideId)
```
Comment (axthosarouris) [inline publication-commons/src/main/java/no/unit/nva/publication/external/services/PersonApiClient.java:51]: The only public method of this class: It returns the CristinUnit Ids for a User.

File: publication-commons/src/main/java/no/unit/nva/publication/external/services/PersonApiClient.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        throws IOException, InterruptedException, ApiGatewayException {
+        var queryUri = createQueryToPersonService(feideId);
+        HttpResponse<String> response = sendQuery(queryUri);
+        checkResponse(response);
+        return extractUserAffiliations(response);
+    }
+
+    private void checkResponse(HttpResponse<String> response) throws BadGatewayException {
```
Comment (axthosarouris) [inline publication-commons/src/main/java/no/unit/nva/publication/external/services/PersonApiClient.java:59]: We cannot handle anything else except for OK

File: publication-commons/src/main/java/no/unit/nva/publication/external/services/PersonApiClient.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        if (!Objects.equals(HttpURLConnection.HTTP_OK,response.statusCode())) {
+            Problem errorReport = generateErrorReport(response);
+            logger.warn(PERSON_SERVICE_RESPONSE + errorReport);
+            throw new BadGatewayException(ERROR_WITH_PERSON_SERVICE_COMMINUCATION);
+        }
+    }
+
+    private Problem generateErrorReport(HttpResponse<String> response) {
```
Comment (axthosarouris) [inline publication-commons/src/main/java/no/unit/nva/publication/external/services/PersonApiClient.java:67]: Log the error as a Zalando problem. It is Json and it has a standard format

File: publication-commons/src/main/java/no/unit/nva/publication/external/services/PersonApiClient.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    private List<URI> extractUserAffiliations(HttpResponse<String> response) throws JsonProcessingException {
+        return extractUserInformation(response)
+            .map(details -> (ArrayNode) details.at(USER_AFFILIATIONS_FIELD))
+            .map(this::toUriList)
+            .orElse(Collections.emptyList());
+    }
+
+    private Optional<ObjectNode> extractUserInformation(HttpResponse<String> response) throws JsonProcessingException {
+        var result = (ArrayNode) dtoObjectMapper.readTree(response.body());
```
Comment (brinxmat) [inline publication-commons/src/main/java/no/unit/nva/publication/external/services/PersonApiClient.java:98]: The response is always a list, right? IIRC, there is a way around this, but perhaps not relevant here

**Review by brinxmat — CHANGES_REQUESTED** [review]

**Review by brinxmat — APPROVED** [review]

**Pull request merged**